File: src/marketAnalysis.ts
import { marketDatasetPath, normalizeStooqSymbol } from './data.ts';
import { readJsonl, type JsonObject } from './storage.ts';

const TRADING_DAYS = 252;

function numberOrNull(value: unknown): number | null {
  const parsed = Number(String(value ?? '').replaceAll(',', ''));
  return Number.isFinite(parsed) ? parsed : null;
}

export type MarketRow = {
  date: string;
  ticker: string;
  provider_symbol: string;
  source: string;
  interval: string;
  open: number | null;
  high: number | null;
  low: number | null;
  close: number;
  volume: number | null;
};

export function marketRows(
  symbol: string,
  options: { base?: string; source?: string; interval?: string; providerSymbol?: string } = {},
): MarketRow[] {
  const source = options.source || 'stooq';
  const interval = options.interval || 'd';
  const resolvedSymbol = source === 'stooq' ? normalizeStooqSymbol(symbol, options.providerSymbol) : (options.providerSymbol || symbol).toLowerCase();
  const records = readJsonl(marketDatasetPath(options.base || 'data', source, resolvedSymbol, interval));
  return records.flatMap((record) => {
    const payload = typeof record.payload === 'object' && record.payload && !Array.isArray(record.payload) ? record.payload as JsonObject : {};
    const close = numberOrNull(payload.close);
    if (close === null) return [];
    return [{
      date: String(record.date ?? ''),
      ticker: String(record.ticker ?? symbol.toUpperCase()).toUpperCase(),
      provider_symbol: String(record.provider_symbol ?? resolvedSymbol),
      source: String(record.source ?? source),
      interval: String(record.interval ?? interval),
      open: numberOrNull(payload.open),
      high: numberOrNull(payload.high),
      low: numberOrNull(payload.low),
      close,
      volume: numberOrNull(payload.volume),
    }];
  }).sort((a, b) => a.date.localeCompare(b.date));
}

function returns(closes: number[]): number[] {
  const values: number[] = [];
  for (let i = 1; i < closes.length; i += 1) {
    const previous = closes[i - 1]!;
    const current = closes[i]!;
    if (previous !== 0) values.push(current / previous - 1);
  }
  return values;
}

function mean(values: number[]): number | null {
  return values.length ? values.reduce((sum, value) => sum + value, 0) / values.length : null;
}

function stddev(values: number[]): number | null {
  if (values.length < 2) return null;
  const average = mean(values)!;
  return Math.sqrt(values.reduce((sum, value) => sum + ((value - average) ** 2), 0) / (values.length - 1));
}

function maxDrawdown(closes: number[]): number | null {
  if (!closes.length) return null;
  let peak = closes[0]!;
  let worst = 0;
  for (const close of closes) {
    peak = Math.max(peak, close);
    if (peak !== 0) worst = Math.min(worst, close / peak - 1);
  }
  return worst;
}

function movingAverage(values: number[], window: number): number | null {
  if (values.length < window) return null;
  return mean(values.slice(-window));
}

function volumeRatio(rows: MarketRow[], window = 20): number | null {
  const volumes = rows.map((row) => row.volume).filter((value): value is number => value !== null);
  if (volumes.length < window || !volumes.at(-1)) return null;
  const average = mean(volumes.slice(-window));
  return average === null || average === 0 ? null : volumes.at(-1)! / average;
}

function regime(totalReturn: number | null, latestClose: number | null, ma20: number | null, ma50: number | null, volatility: number | null): string {
  if (latestClose === null) return 'no-price-data';
  if (ma20 !== null && ma50 !== null && latestClose > ma20 && ma20 > ma50) return 'trend-up';
  if (ma20 !== null && ma50 !== null && latestClose < ma20 && ma20 < ma50) return 'trend-down';
  if (volatility !== null && volatility > 0.04) return 'high-volatility';
  if (totalReturn !== null && Math.abs(totalReturn) < 0.02) return 'range-bound';
  return 'watch';
}

export function marketStats(
  symbol: string,
  options: { base?: string; source?: string; interval?: string; providerSymbol?: string } = {},
): JsonObject {
  const source = options.source || 'stooq';
  const interval = options.interval || 'd';
  const rows = marketRows(symbol, { ...options, source, interval });
  if (!rows.length) {
    return {
      ok: false,
      ticker: symbol.toUpperCase(),
      source,
      interval,
      rows: 0,
      error: 'no market dataset found; run data download first',
      next_command: `data download ${symbol.toUpperCase()}`,
    };
  }
  const closes = rows.map((row) => row.close);
  const rowReturns = returns(closes);
  const latestClose = closes.at(-1)!;
  const firstClose = closes[0]!;
  const totalReturn = firstClose === 0 ? null : latestClose / firstClose - 1;
  const volatility = stddev(rowReturns);
  const ma20 = movingAverage(closes, 20);
  const ma50 = movingAverage(closes, 50);
  return {
    ok: true,
    ticker: rows.at(-1)!.ticker,
    provider_symbol: rows.at(-1)!.provider_symbol,
    source,
    interval,
    rows: rows.length,
    start_date: rows[0]!.date,
    end_date: rows.at(-1)!.date,
    latest_close: latestClose,
    total_return: totalReturn,
    average_return: mean(rowReturns),
    volatility,
    annualized_volatility: volatility === null ? null : volatility * Math.sqrt(TRADING_DAYS),
    max_drawdown: maxDrawdown(closes),
    best_return: rowReturns.length ? Math.max(...rowReturns) : null,
    worst_return: rowReturns.length ? Math.min(...rowReturns) : null,
    moving_average_20: ma20,
    moving_average_50: ma50,
    latest_volume: rows.at(-1)!.volume,
    volume_ratio_20: volumeRatio(rows, 20),
    regime: regime(totalReturn, latestClose, ma20, ma50, volatility),
    readiness: {
      basic_stats: rows.length >= 2,
      moving_average_20: rows.length >= 20,
      moving_average_50: rows.length >= 50,
      backtest_ready: rows.length >= 60,
    },
  };
}

File: tossquant_cli/market_analysis.py
from __future__ import annotations

import math
from pathlib import Path
from statistics import fmean
from typing import Any

from .data import market_dataset_path, normalize_stooq_symbol
from .storage import read_jsonl

TRADING_DAYS = 252


def _number(value: Any) -> float | None:
    try:
        parsed = float(str(value).replace(",", ""))
    except (TypeError, ValueError):
        return None
    return parsed if math.isfinite(parsed) else None


def market_rows(
    symbol: str,
    *,
    base: str | Path | None = None,
    source: str = "stooq",
    interval: str = "d",
    provider_symbol: str | None = None,
) -> list[dict[str, Any]]:
    resolved_symbol = normalize_stooq_symbol(symbol, provider_symbol) if source == "stooq" else (provider_symbol or symbol).lower()
    path = market_dataset_path(base, source, resolved_symbol, interval)
    records = read_jsonl(path)
    rows = []
    for record in records:
        payload = record.get("payload") if isinstance(record, dict) else None
        payload = payload if isinstance(payload, dict) else {}
        close = _number(payload.get("close"))
        if close is None:
            continue
        rows.append(
            {
                "date": record.get("date"),
                "ticker": record.get("ticker", symbol.upper()),
                "provider_symbol": record.get("provider_symbol", resolved_symbol),
                "source": record.get("source", source),
                "interval": record.get("interval", interval),
                "open": _number(payload.get("open")),
                "high": _number(payload.get("high")),
                "low": _number(payload.get("low")),
                "close": close,
                "volume": _number(payload.get("volume")),
            }
        )
    return sorted(rows, key=lambda row: str(row.get("date") or ""))


def _returns(closes: list[float]) -> list[float]:
    values = []
    for previous, current in zip(closes, closes[1:]):
        if previous == 0:
            continue
        values.append(current / previous - 1.0)
    return values


def _stddev(values: list[float]) -> float | None:
    if len(values) < 2:
        return None
    mean = fmean(values)
    return math.sqrt(sum((value - mean) ** 2 for value in values) / (len(values) - 1))


def _max_drawdown(closes: list[float]) -> float | None:
    if not closes:
        return None
    peak = closes[0]
    worst = 0.0
    for close in closes:
        peak = max(peak, close)
        if peak:
            worst = min(worst, close / peak - 1.0)
    return worst


def _moving_average(values: list[float], window: int) -> float | None:
    if len(values) < window:
        return None
    return fmean(values[-window:])


def _volume_ratio(rows: list[dict[str, Any]], window: int = 20) -> float | None:
    volumes = [row["volume"] for row in rows if row.get("volume") is not None]
    if len(volumes) < window or not volumes[-1]:
        return None
    average = fmean(volumes[-window:])
    return None if average == 0 else volumes[-1] / average


def _regime(total_return: float | None, latest_close: float | None, ma20: float | None, ma50: float | None, volatility: float | None) -> str:
    if latest_close is None:
        return "no-price-data"
    if ma20 is not None and ma50 is not None and latest_close > ma20 > ma50:
        return "trend-up"
    if ma20 is not None and ma50 is not None and latest_close < ma20 < ma50:
        return "trend-down"
    if volatility is not None and volatility > 0.04:
        return "high-volatility"
    if total_return is not None and abs(total_return) < 0.02:
        return "range-bound"
    return "watch"


def market_stats(
    symbol: str,
    *,
    base: str | Path | None = None,
    source: str = "stooq",
    interval: str = "d",
    provider_symbol: str | None = None,
) -> dict[str, Any]:
    rows = market_rows(symbol, base=base, source=source, interval=interval, provider_symbol=provider_symbol)
    if not rows:
        return {
            "ok": False,
            "ticker": symbol.upper(),
            "source": source,
            "interval": interval,
            "rows": 0,
            "error": "no market dataset found; run data download first",
            "next_command": f"data download {symbol.upper()}",
        }

    closes = [float(row["close"]) for row in rows]
    returns = _returns(closes)
    latest_close = closes[-1]
    first_close = closes[0]
    total_return = None if first_close == 0 else latest_close / first_close - 1.0
    volatility = _stddev(returns)
    annualized_volatility = None if volatility is None else volatility * math.sqrt(TRADING_DAYS)
    ma20 = _moving_average(closes, 20)
    ma50 = _moving_average(closes, 50)
    volume_ratio_20 = _volume_ratio(rows, 20)
    return {
        "ok": True,
        "ticker": str(rows[-1].get("ticker") or symbol.upper()).upper(),
        "provider_symbol": rows[-1].get("provider_symbol"),
        "source": source,
        "interval": interval,
        "rows": len(rows),
        "start_date": rows[0].get("date"),
        "end_date": rows[-1].get("date"),
        "latest_close": latest_close,
        "total_return": total_return,
        "average_return": fmean(returns) if returns else None,
        "volatility": volatility,
        "annualized_volatility": annualized_volatility,
        "max_drawdown": _max_drawdown(closes),
        "best_return": max(returns) if returns else None,
        "worst_return": min(returns) if returns else None,
        "moving_average_20": ma20,
        "moving_average_50": ma50,
        "latest_volume": rows[-1].get("volume"),
        "volume_ratio_20": volume_ratio_20,
        "regime": _regime(total_return, latest_close, ma20, ma50, volatility),
        "readiness": {
            "basic_stats": len(rows) >= 2,
            "moving_average_20": len(rows) >= 20,
            "moving_average_50": len(rows) >= 50,
            "backtest_ready": len(rows) >= 60,
        },
    }
